GitMaterial Authentication failed for the second user who using same repository

##### Issue Type

<!--- Please specify the issue type to help us categorize the issue, mention any one of the below types -->

- Bug Report

##### Summary

If two git materials are configured in two pipelines with same git repo and different username + accesstoken, the second user will failed to checkout.

##### Environment

<!--- 
Describe the environment in which you encountered this issue.
Details of the environment GoCD is running in can be fetched from the GoCD support api - http://<go-server-host>/go/api/support.
If you are pasting plain text, please surround it with 3 backticks on each side
-->
```plain
"Go Server Information": {
    "Version": "20.6.0 (12005-12860aac6351e2a353728c7d7913f34d741c63e0)"
  }，
"OS Information": {
    "OS Name": "Linux",
    "OS Version": "3.10.0-862.el7.x86_64",
    "System Architecture": "amd64",
    "Available Processors": 24,
    "Average System Load": 0.67
  },
```


###### Basic environment details

<!--- We recommend providing the for us to reproduce the issue quicker -->

* Go Version: `20.6.0 (12005-12860aac6351e2a353728c7d7913f34d741c63e0)`
* JAVA Version: `13.0.2`
* OS: `Linux 3.10.0-862.el7.x86_64`

###### Additional Environment Details

<!-- More environment details captured from the support API or other sources can be shared here -->

##### Steps to Reproduce

<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug include code to reproduce, if relevant -->
1. create two pipeline with git credential `https://gitlab.somehost/someproject.git`, username and password(gitlab accesstoken) are provided in advanced settings. 
2. the username/accesstokens of those git material are belongs to two person. now call them userA and userB.
3. While click `test connection` in `Edit material - Git`, both repositories says: "Connection OK".
4. trigger the first pipeline mannually, later it will success.
5. trigger the second pipeline mannually, later it will output `Authentication failed`

##### Expected Results
The second pipeline should works as the first one.

##### Actual Results
The second pipeline build failed with message `Authentication failed` and `Failed to run git clone command`

##### Possible Fix
<!--- Not obligatory, but suggest a fix or reason for the bug -->
I have read some source code, the access token are stored correctly in cruise file.The material editing page uses it too.

The method `com.thoughtworks.go.config.materials.git.GitMaterial#appendCriteria` is like:
```
        parameters.put(ScmMaterialConfig.URL, url.originalArgument());
        parameters.put("branch", refSpecOrBranch);
```
the `generateFingerprintFromCriteria` is using `type`, `url`, `branch`, it will using the first pipeline's material.

It seems that is agent JVM, the `com.thoughtworks.go.config.materials.ScmMaterial`:
```
    protected String userName; //the userName is always UserA.
    protected String password;
```

I cannot figure out why `GitMaterial` failed to checkout userB's pipeline.


##### Log snippets
To see command line logs, provide:
logback-agent-include.xml
```
<logger name="com.thoughtworks.go.util.command.CommandLine" level="DEBUG" >
     <appender-ref ref="commandLineLog"/>
</logger>
```


Agentlog
```
2021-01-20 17:15:51,543 DEBUG [scheduler-3] CommandLine:350 - Running git config remote.origin.url
2021-01-20 17:15:51,681 DEBUG [scheduler-3] CommandLine:362 - Output: 
https://UserB:UserB'Accesstoken@gitlab.host/project.git
2021-01-20 17:15:51,778 DEBUG [scheduler-3] CommandLine:368 - Running git clone --branch deploy https://UserA:******@gitlab.host/project.git /var/lib/go-agent/pipelines/HandyCreated
2021-01-20 17:15:52,164 ERROR [scheduler-3] DefaultGoPublisher:137 - Failed to run git clone command
java.lang.RuntimeException: Failed to run git clone command
        at com.thoughtworks.go.util.ExceptionUtils.bomb(ExceptionUtils.java:37)
        at com.thoughtworks.go.config.materials.git.GitMaterial.updateTo(GitMaterial.java:157)
        at com.thoughtworks.go.domain.MaterialRevision.updateTo(MaterialRevision.java:147)
        at com.thoughtworks.go.domain.materials.AbstractMaterialAgent.prepare(AbstractMaterialAgent.java:43)
......
Caused by: java.lang.Exception: Failed to run git clone command
        at com.thoughtworks.go.util.ExceptionUtils.bombIfFailedToRunCommandLine(ExceptionUtils.java:51)
        at com.thoughtworks.go.config.materials.git.GitMaterial.git(GitMaterial.java:384)
        at com.thoughtworks.go.config.materials.git.GitMaterial.updateTo(GitMaterial.java:151)
        ... 20 common frames omitted

```

```
[go] Start to update materials.

[go] Start updating files at revision 65d9a85e92e8284857409fe62fcef2d80fb9211c from https://gitlab.<host>.com/<project>.git
Cloning into '/var/lib/go-agent/pipelines/HandyCreated'...
STDERR: fatal: Authentication failed for 'https://<userA>:******@<host>/<project>.git/'
Failed to run git clone command
```

##### Code snippets/Screenshots
![image](https://user-images.githubusercontent.com/8142133/105148039-e00c4d00-5b3c-11eb-8a2a-0eaf35b3646c.png)
![image](https://user-images.githubusercontent.com/8142133/105148223-14800900-5b3d-11eb-9f85-409d08834091.png)


##### Any other info
<!--- Provide any other information you would like to share to explain this issue more better -->

--- a/common/src/test/java/com/thoughtworks/go/domain/materials/AbstractMaterialTest.java
+++ b/common/src/test/java/com/thoughtworks/go/domain/materials/AbstractMaterialTest.java
@@ -18,6 +18,8 @@ package com.thoughtworks.go.domain.materials;
 import com.thoughtworks.go.config.PipelineConfig;
 import com.thoughtworks.go.config.materials.AbstractMaterial;
 import com.thoughtworks.go.config.materials.SubprocessExecutionContext;
+import com.thoughtworks.go.config.materials.svn.SvnMaterial;
+import com.thoughtworks.go.config.materials.svn.SvnMaterialConfig;
 import com.thoughtworks.go.domain.MaterialInstance;
 import com.thoughtworks.go.domain.MaterialRevision;
 import com.thoughtworks.go.util.command.ConsoleOutputStreamConsumer;
@@ -30,7 +32,7 @@ import java.util.Map;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatCode;
-import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.*;
 
 public class AbstractMaterialTest {
 
@@ -206,5 +208,28 @@ public class AbstractMaterialTest {
         assertThatCode(() -> material.checkConnection(executionContext))
                 .isInstanceOf(UnsupportedOperationException.class)
                 .hasMessage("'checkConnection' cannot be performed on material of type name");
+    }
+
+    @Test
+    public void updateConfigShouldIgnoreNonPasswordAwareConfig() {
+        MaterialConfig config = mock(MaterialConfig.class);
+        SvnMaterial passwordAwareMaterial = new SvnMaterial("url", "username", "password", false);
+        passwordAwareMaterial.updateFromConfig(config);
+        verifyNoInteractions(config);
+    }
+
+    @Test
+    public void updateConfigShouldTakeUserPassFromPasswordAwareConfig() {
+        SvnMaterialConfig config = mock(SvnMaterialConfig.class);
+        when(config.getUserName()).thenReturn("newUsername");
+        when(config.getPassword()).thenReturn("newPassword");
+        SvnMaterial passwordAwareMaterial = new SvnMaterial("url", "username", "password", false);
+        passwordAwareMaterial.updateFromConfig(config);
+
+        assertThat(passwordAwareMaterial.getUserName()).isEqualTo("newUsername");
+        assertThat(passwordAwareMaterial.getPassword()).isEqualTo("newPassword");
+
+        verify(config).getUserName();
+        verify(config).getPassword();
     }
 }

--- a/config/config-api/src/main/java/com/thoughtworks/go/config/materials/AbstractMaterialConfig.java
+++ b/config/config-api/src/main/java/com/thoughtworks/go/config/materials/AbstractMaterialConfig.java
@@ -121,8 +121,8 @@ public abstract class AbstractMaterialConfig implements MaterialConfig, ParamsAt
     public String getTruncatedDisplayName() {
         String displayName = getDisplayName();
         if (displayName.length() > TRUNCATED_NAME_MAX_LENGTH) {
-            StringBuffer buffer = new StringBuffer();
-            buffer.append(displayName.substring(0, TRUNCATED_NAME_MAX_LENGTH / 2));
+            StringBuilder buffer = new StringBuilder();
+            buffer.append(displayName, 0, TRUNCATED_NAME_MAX_LENGTH / 2);
             buffer.append("...");
             buffer.append(displayName.substring(displayName.length() - TRUNCATED_NAME_MAX_LENGTH / 2));
             displayName = buffer.toString();

--- a/config/config-api/src/main/java/com/thoughtworks/go/config/materials/PasswordAwareMaterial.java
+++ b/config/config-api/src/main/java/com/thoughtworks/go/config/materials/PasswordAwareMaterial.java
@@ -16,6 +16,8 @@
 package com.thoughtworks.go.config.materials;
 
 public interface PasswordAwareMaterial {
+    void setUserName(String userName);
+    String getUserName();
     void setPassword(String password);
     String getPassword();
 }

--- a/config/config-api/src/main/java/com/thoughtworks/go/domain/materials/TestingMaterialConfig.java
+++ b/config/config-api/src/main/java/com/thoughtworks/go/domain/materials/TestingMaterialConfig.java
@@ -17,17 +17,10 @@ package com.thoughtworks.go.domain.materials;
 
 import com.thoughtworks.go.config.ValidationContext;
 import com.thoughtworks.go.config.materials.ScmMaterialConfig;
-import org.joda.time.DateTime;
 
-import java.util.Date;
 import java.util.Map;
 
-public class TestingMaterialConfig extends ScmMaterialConfig {
-    public static final Date TWO_DAYS_AGO_CHECKIN = new DateTime().minusDays(2).toDate();
-
-    public static final String MOD_TYPE = "svn";
-    public static final String MOD_REVISION = "98";
-
+public class TestingMaterialConfig extends ScmMaterialConfig{
     private static final String TYPE = "TestingMaterial";
 
     private String url;

--- a/domain/src/main/java/com/thoughtworks/go/config/materials/AbstractMaterial.java
+++ b/domain/src/main/java/com/thoughtworks/go/config/materials/AbstractMaterial.java
@@ -226,6 +226,7 @@ public abstract class AbstractMaterial extends PersistentObject implements Mater
     public void updateFromConfig(MaterialConfig materialConfig) {
         if (materialConfig instanceof PasswordAwareMaterial) {
             PasswordAwareMaterial passwordConfig = (PasswordAwareMaterial) materialConfig;
+            ((PasswordAwareMaterial) this).setUserName(passwordConfig.getUserName());
             ((PasswordAwareMaterial) this).setPassword(passwordConfig.getPassword());
         }
     }

--- a/server/src/test-fast/java/com/thoughtworks/go/server/service/ScheduledPipelineLoaderTest.java
+++ b/server/src/test-fast/java/com/thoughtworks/go/server/service/ScheduledPipelineLoaderTest.java
@@ -1,0 +1,80 @@
+/*
+ * Copyright 2022 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.server.service;
+
+import com.thoughtworks.go.config.materials.PasswordAwareMaterial;
+import com.thoughtworks.go.config.materials.git.GitMaterialConfig;
+import com.thoughtworks.go.domain.Pipeline;
+import com.thoughtworks.go.helper.MaterialConfigsMother;
+import com.thoughtworks.go.server.dao.PipelineSqlMapDao;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoSettings;
+
+import static com.thoughtworks.go.helper.ConfigFileFixture.configWith;
+import static com.thoughtworks.go.helper.GoConfigMother.createPipelineConfigWithMaterialConfig;
+import static com.thoughtworks.go.helper.PipelineMother.preparing;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.when;
+
+@MockitoSettings
+class ScheduledPipelineLoaderTest {
+
+    @Mock
+    private PipelineSqlMapDao pipelineDao;
+    @Mock
+    private GoConfigService goConfigService;
+    @InjectMocks
+    private MaterialExpansionService materialExpansionService;
+
+    private ScheduledPipelineLoader loader;
+
+    @BeforeEach
+    public void setUp() {
+        loader = new ScheduledPipelineLoader(pipelineDao, goConfigService, null, null, null, materialExpansionService, null);
+    }
+
+    @Test
+    public void materialUserPassShouldBeTakenFromConfigOnLoad() {
+
+        // De-duplication of materials may lead to the triggering material having different credentials to the
+        // configured material specific to this pipeline.
+        GitMaterialConfig triggeredMaterial = MaterialConfigsMother.git("https://url", "userNameFromTrigger", null);
+        GitMaterialConfig configuredMaterial = MaterialConfigsMother.git("https://url", "userFromConfig", "passFromConfig");
+
+        assertThat(triggeredMaterial.getUserName()).isNotEqualTo(configuredMaterial.getUserName());
+        assertThat(triggeredMaterial.getPassword()).isNotEqualTo(configuredMaterial.getPassword());
+
+        when(pipelineDao.pipelineWithMaterialsAndModsByBuildId(1))
+                .thenReturn(preparing(createPipelineConfigWithMaterialConfig(triggeredMaterial)));
+
+        when(goConfigService.getCurrentConfig())
+                .thenReturn(configWith(createPipelineConfigWithMaterialConfig(configuredMaterial)));
+
+        Pipeline loadedPipeline = loader.pipelineWithPasswordAwareBuildCauseByBuildId(1);
+
+        assertThat(loadedPipeline.getMaterialRevisions())
+                .singleElement()
+                .satisfies(rev -> assertThat(rev.getMaterial().config())
+                        .isInstanceOfSatisfying(PasswordAwareMaterial.class, mat -> {
+                            assertThat(mat.getUserName()).isEqualTo("userFromConfig");
+                            assertThat(mat.getPassword()).isEqualTo("passFromConfig");
+                        }));
+    }
+}

